File: tools/idd-tool/Installation/FileInstaller.cs
internal sealed class FileInstaller
{
    public void Copy(IReadOnlyList<PlannedFile> files, string destinationRoot, bool force)
    {
        var conflicts = files
            .Where(file =>
            {
                var destination = Path.Combine(destinationRoot, file.RelativePath);
                return File.Exists(destination) &&
                    !StringComparer.Ordinal.Equals(FileHasher.Sha256(File.ReadAllBytes(destination)), file.Hash);
            })
            .Select(file => file.RelativePath)
            .ToArray();

        if (conflicts.Length > 0 && !force)
        {
            throw new ToolException(string.Join(Environment.NewLine, conflicts.Select(path => $"File already exists: {path}")) +
                Environment.NewLine +
                "Use --force to overwrite.");
        }

        foreach (var file in files)
        {
            var destination = Path.Combine(destinationRoot, file.RelativePath);
            Directory.CreateDirectory(Path.GetDirectoryName(destination)!);
            File.WriteAllBytes(destination, file.Content);
        }
    }
}

File: npm/lib/file-installer.js
const fs = require("fs");
const path = require("path");
const { fail } = require("./errors");
const { sha256 } = require("./hash");

function copyPlannedFiles(files, destinationRoot, force) {
  const conflicts = files
    .filter((file) => {
      const destination = path.join(destinationRoot, file.relativePath);
      return fs.existsSync(destination) && sha256(fs.readFileSync(destination)) !== file.hash;
    })
    .map((file) => file.relativePath);

  if (conflicts.length > 0 && !force) {
    for (const relativePath of conflicts) {
      console.error(`File already exists: ${relativePath}`);
    }

    fail("Use --force to overwrite.");
  }

  for (const file of files) {
    const destination = path.join(destinationRoot, file.relativePath);
    fs.mkdirSync(path.dirname(destination), { recursive: true });
    if (file.content) {
      fs.writeFileSync(destination, file.content);
    } else {
      fs.copyFileSync(file.sourcePath, destination);
    }
  }
}

module.exports = { copyPlannedFiles };
